### PR TITLE
Add pipeline framework to make parallel processing simpler

### PR DESCRIPTION
**Pull request by ajsutton**
## PR description
Adds a new pipeline module to provide a simple way of chaining processing that needs to happen in parallel.  This will wind up being used in the `WorldStateDownloader` to improve its performance.

World state downloading will come out something like:
```
final Pipeline pipeline =
          createPipelineFrom(new TaskQueueIterator(newDownloadState), persistenceQueueCapacity)
              .thenProcess(this::checkForLocalData)
              .inBatches(hashCountPerRequest)
              .thenProcessAsync(
                  requestTasks -> sendRequestForData(requestTasks, header, newDownloadState),
                  maxOutstandingRequests)
              .thenProcess(tasks -> persistBatch(tasks, header, newDownloadState))
              .thenFlatMap(List::stream, persistenceQueueCapacity)
              .thenProcessInParallel(task -> enqueueChildren(task, header, newDownloadState), 3)
              .andFinishWith(task -> markAsCompleteOrFailed(header, newDownloadState, task));
```

### CODE REVIEW

File: services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipe.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+   * Get the number of items that can be queued inside this pipe.
+   *
+   * @return the pipe's capacity.
+   */
+  public int getCapacity() {
+    return capacity;
+  }
+
+  @Override
+  public boolean hasRemainingCapacity() {
```
Comment (mbaxter) [inline services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipe.java]: If the pipe is not open, should this return 0 ??

File: services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/OutputPipe.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.services.pipeline;
+
+/**
+ * The end of the pipe that stages write their output to.
+ *
+ * @param <T> the type of output.
+ */
+public interface OutputPipe<T> {
```
[the remark this answers is not in the document]
Comment (ajsutton) [reply services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/OutputPipe.java]: It depends which way you look at it.  The place where pipes are actually held is mostly inside a stage where it reads inputs from it's `InputPipe` and writes outputs to it's `OutputPipe`. Admittedly it is a bit weird to think of a pipe as being made up of an output which leads to an input but most of the time we're looking at the pipe from the point of view of a stage which the naming works.

When I tried renaming I inevitably wound up with APIs like:
`void processNextInput(PipeOutput inputPipe, PipeInput outputPipe)`

So I've side-stepped the issue a bit and used `ReadPipe` and `WritePipe`. The variable names are unchanged so a stage has an API like:
`void processNextInput(ReadPipe inputPipe, WritePipe outputPipe)`

File: services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipeline.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+   * without the completion stage then marking the future successful before we finish the abort
+   * process and mark it as exceptionally completed. We can't just use synchronized because it winds
+   * up being the same thread coming in via a callback so already has the lock.
+   */
+  private final AtomicBoolean completing = new AtomicBoolean(false);
+
+  private final CompletableFuture<Void> overallFuture = new CompletableFuture<>();
+  private volatile List<Future<?>> futures;
+
+  public Pipeline(
```
[the remark this answers is not in the document]
Comment (ajsutton) [reply services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipeline.java]: 👍

File: services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/PipelineBuilder.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+   *
+   * <p>The output buffer size is reduced to <code>bufferSize / maximumBatchSize + 1</code>.
+   *
+   * @param stageName the name of this stage. Used as the label for the output count metric.
+   * @param maximumBatchSize the maximum number of items to include in a batch.
+   * @return a {@link PipelineBuilder} ready to extend the pipeline with additional stages.
+   */
+  public PipelineBuilder<List<T>> inBatches(final String stageName, final int maximumBatchSize) {
+    return addStage(
+        new BatchingProcessor<>(maximumBatchSize), bufferSize / maximumBatchSize + 1, stageName);
```
[the remark this answers is not in the document]
Comment (ajsutton) [reply services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/PipelineBuilder.java]: Seems reasonable.

File: services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipe.java (hunk 1 above)
Comment (ajsutton) [reply services/pipeline/src/main/java/tech/pegasys/pantheon/services/pipeline/Pipe.java]: Also seems reasonable.

**Pull request merged**